src/matcha_ml/services/global_config_service.py

```diff
@@ -0,0 +1,139 @@
+"""Global config service for creating and modifying a users global config files."""
+import os
+import uuid
+from typing import Any, Dict, Optional
+
+import yaml
+
+from matcha_ml.errors import MatchaPermissionError
+
+
+class GlobalConfigurationService:
+    """A Global Config Service for interacting and updating a users global config file."""
+
+    _instance: Optional["GlobalConfigurationService"] = None
+    _user_id: Optional[str] = None
+    _analytics_opt_out: bool = False
+
+    def __new__(cls) -> "GlobalConfigurationService":
+        """Singleton class definition.
+
+        Returns:
+            GlobalConfigurationService: Already existing initialised object, otherwise a new singleton object
+        """
+        if cls._instance is None:
+            cls._instance = super().__new__(cls)
+            # Check if config.yaml file exists and read in variables to the class
+            if os.path.exists(str(cls._instance.default_config_file_path)):
+                cls._instance._read_global_config(
+                    str(cls._instance.default_config_file_path)
+                )
+            else:
+                cls._instance._create_global_config(
+                    config_file_path=str(cls._instance.default_config_file_path)
+                )
+
+        return cls._instance
+
+    def _read_global_config(self, config_file_path: str) -> None:
+        """Reads the global config yaml file.
+
+        Args:
+            config_file_path (str): Path to users global config file
+        """
+        with open(config_file_path) as file:
+            yaml_data = yaml.safe_load(file)
+            self._user_id = yaml_data.get("user_id")
+            self._analytics_opt_out = yaml_data.get("analytics_opt_out") == "True"
+
+    def _create_global_config(self, config_file_path: str) -> None:
+        """Creates a new global config yaml file.
+
+        Args:
+            config_file_path (str): Path to users global config file
+        """
+        # Generate a new unique user ID
+        data = {
+            "user_id": self.user_id,
+            "analytics_opt_out": self._analytics_opt_out,
+        }
+
+        # Create the '.matcha-ml' config directory
+        try:
+            os.makedirs(os.path.dirname(config_file_path), exist_ok=True)
+        except PermissionError:
+            raise MatchaPermissionError(
+                f"Error - You do not have permission to write the configuration. Check if you have write permissions for '{config_file_path}'"
+            )
+
+        # Create config file and populate with the current class variables
+        with open(config_file_path, "w") as file:
+            yaml.dump(data, file)
+
+    def _update_global_config(self, config_file_path: str) -> None:
```

Comment:
Why is `config_file_path` passed as a parameter here and rest of the methods? We can directly use `self.default_config_file_path` variable instead of this parameter.